kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -594,17 +619,27 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
         cancelImpl(parentJob)
     }
 
-    // Child was cancelled with cause
-    // It is overridden in supervisor implementations to ignore child cancellation
-    public open fun childCancelled(cause: Throwable): Boolean =
-        cancelImpl(cause) && handlesException
+    /**
+     * Returns `true` if job should cancel itself on child [CancellationException].
+     */
+    public open fun cancelOnChildCancellation(cause: CancellationException) = false
+
+    /**
+     * Child was cancelled with a cause.
+     * In this method parent decides whether it cancels itself (e.g. on a critical failure) and
+     * whether it handles the exception of the child.
+     * It is overridden in supervisor implementations to completely ignore any child cancellation.
+     */
+    public open fun childCancelled(cause: Throwable): Boolean {
+        if (cause is CancellationException && !cancelOnChildCancellation(cause)) return true
+        return cancelImpl(cause) && handlesException
+    }
```

Comment:
I suggest to replace it with the following code to make "default" cancellation logic clearer and to fix a bug with `produce` coroutine (failure of a child shall immediately cancel the channel because it is doomed to fail and should not wait for main coroutine's failure):
```
    public open fun childCancelled(cause: Throwable): Boolean {
        if (cause is CancellationException) return true
        return cancelInternal(cause)
    }
```